### ISSUE TITLE
system runs of memory at index creation time

### ISSUE THREAD

**Issue by rationalga** [lancedb/lancedb#767]
Hi,
I have inserted 20 million embeddings (768 dimensions each) in a table. The insertion took around 1 hour and 48 minutes which is great. The memory consumption during insertion was around 21 GB. 
When I tried to create index as below

`table.create_index(metric="cosine", num_partitions=16384, num_sub_vectors=64)
`
After a day the system runs out of memory and indexing fails. The average memory consumption initally while indexing was running was around 20.5GB.
Can anyone please guide that
1. Why is the training step of the IVFPQ implementation consuming such a significant amount of memory, despite being a disk-based process?
2. How can I ensure that both the training and searching processes are primarily utilizing SSD storage, rather than being limited by RAM capacity?
Thank you very much.

**Comment by chebbyChefNEQ** [comment]
Hi,

The index write buffers two 10M row partitions at once before flushing. This is why the memory consumion is high. I'll expose these settings so you can tune down the memory useage

**Comment by rationalga** [comment]
Thank you for you replies.
yes the table.add() took 1 hour and 48 minutes. The  table.create_index never finished. I left it running for 24 hours and then it eventually crashed.

**Comment by rationalga** [comment]
As an additional questions, is there a plan of exposing IVF_FLAT  index in python? It would be great as the users would have the choice whether to use compression or not.

**Comment by changhiskhan** [comment]
> Thank you for you replies. yes the table.add() took 1 hour and 48 minutes. The table.create_index never finished. I left it running for 24 hours and then it eventually crashed.

are you adding one row at a time? If you can add in batches of like 10K or 100K that should be super fast. If you add one row at a time, it's going to create too many small files which makes ingestion slower AND the indexing would be slow as well

**Comment by rationalga** [comment]
I am adding with a batch size of 10K. I could try 100K. The 1 hour 48 minutes also includes some small preprocessing time on the batch.
batch_size =10000
full_ds.map(lambda batch: process_and_insert_batch(batch, table), batched=True, batch_size=batch_size)

**Comment by changhiskhan** [comment]
Just writing 20M rows should be O(sec).
1. what's in `process_and_insert_batch` ? 
2. Is the embedding generation time included in the 1h48min?
3. what do the other columns look like?

**Comment by rationalga** [comment]
You are right, as I am only inserting data and indexing is done as a separate step then insertion should be fast. However, I think my insertion loop is not very optimised. I was mainly trying to ascertain how lancdeDB is effective at using disk-resident indexes. I want to choose between Vanilla DiskANN implementation and LancDB. 
Anywhere, here is relevant part of the code, please let me know if there is an obvious mistake that I am making otherwise I can try to create a small reproducible code. 
// This is the schema
class MethodRow(LanceModel):
    hexsha: str
    ext: str
    lang: str
    max_stars_repo_path: str
    max_stars_repo_name: str
    code: str
    start_line: int  
    end_line: int   
    vector: Vector(768)  
    
// this is how I restructre the data from the dataset and then add into table

def process_and_insert_batch(batch, table):
    processed_batch = []

    # Determine the number of rows in the batch
    num_rows_in_batch = len(batch['hexsha'])  # Assuming 'hexsha' is always present

    for i in range(num_rows_in_batch):
        hexsha = batch['hexsha'][i]
        ext = batch['ext'][i]
        lang = batch['lang'][i]
        max_stars_repo_path = batch['max_stars_repo_path'][i]
        max_stars_repo_name = batch['max_stars_repo_name'][i]

        # each method within methods has corresponding embedding at same index
        for method, embedding in zip(batch['methods'][i], batch['embeddings'][i]):
            start_line, end_line = method['lines']

            method_data = {
                "hexsha": hexsha,
                "ext": ext,
                "lang": lang,
                "max_stars_repo_path": max_stars_repo_path,
                "max_stars_repo_name": max_stars_repo_name,
                "code": method['body'],
                "start_line": start_line,
                "end_line": end_line,
                "vector": embedding
            }

            processed_batch.append(MethodRow(**method_data).model_dump())
      
    table.add(processed_batch)

And this is how I process the dataset
 #full_ds is huggingface dataset
  full_ds.map(lambda batch: process_and_insert_batch(batch, table), batched=True, batch_size=batch_size)

**Comment by changhiskhan** [comment]
> You are right, as I am only inserting data and indexing is done as a separate step then insertion should be fast. However, I think my insertion loop is not very optimised. I was mainly trying to ascertain how lancdeDB is effective at using disk-resident indexes. I want to choose between Vanilla DiskANN implementation and LancDB. Anywhere, here is relevant part of the code, please let me know if there is an obvious mistake that I am making otherwise I can try to create a small reproducible code. // This is the schema class MethodRow(LanceModel): hexsha: str ext: str lang: str max_stars_repo_path: str max_stars_repo_name: str code: str start_line: int end_line: int vector: Vector(768)
> 
> // this is how I restructre the data from the dataset and then add into table
> 
> def process_and_insert_batch(batch, table): processed_batch = []
> 
> ```
> # Determine the number of rows in the batch
> num_rows_in_batch = len(batch['hexsha'])  # Assuming 'hexsha' is always present
> 
> for i in range(num_rows_in_batch):
>     hexsha = batch['hexsha'][i]
>     ext = batch['ext'][i]
>     lang = batch['lang'][i]
>     max_stars_repo_path = batch['max_stars_repo_path'][i]
>     max_stars_repo_name = batch['max_stars_repo_name'][i]
> 
>     # each method within methods has corresponding embedding at same index
>     for method, embedding in zip(batch['methods'][i], batch['embeddings'][i]):
>         start_line, end_line = method['lines']
> 
>         method_data = {
>             "hexsha": hexsha,
>             "ext": ext,
>             "lang": lang,
>             "max_stars_repo_path": max_stars_repo_path,
>             "max_stars_repo_name": max_stars_repo_name,
>             "code": method['body'],
>             "start_line": start_line,
>             "end_line": end_line,
>             "vector": embedding
>         }
> 
>         processed_batch.append(MethodRow(**method_data).model_dump())
>   
> table.add(processed_batch)
> ```
> 
> And this is how I process the dataset #full_ds is huggingface dataset full_ds.map(lambda batch: process_and_insert_batch(batch, table), batched=True, batch_size=batch_size)

ok so each `table.add` call is inserting `number of rows in the batch * average number of methods per row in the batch` number of vectors into the table. 

You should just be able pass in the Pydantic instances directly into `table.add` and skip the `.model_dump()` call. Are you on pydantic v2? Performance is hugely different between v1 and v2.

Otherwise I don't see anything obvious in this snippet.

Just to clarify, are you saying the 1hr48min *includes* the embedding generation? If so we could try parallelizing that first to shrink this e2e time down?

Btw on indexing, if you have access to a GPU, we support GPU acceleration for index generation, should be much faster (either CUDA or MPS).

**Comment by rationalga** [comment]
> You should just be able pass in the Pydantic instances directly into table.add and skip the .model_dump() call. Are you on pydantic v2? Performance is hugely different between v1 and v2.

Yes, I am using pydantic-2.5.2. Yes I removed model_dump() and it still works. Thanks.

> Just to clarify, are you saying the 1hr48min includes the embedding generation? If so we could try parallelizing that first to shrink this e2e time down?

No, embeddings are precomputed. We just load them from the dataset columns and add them to the table.

> Btw on indexing, if you have access to a GPU, we support GPU acceleration for index generation, should be much faster (either CUDA or MPS).

Oh I see, i'll try that as well. I think I could also reduce the num_partitions and  num_sub_vectors to smaller values to increase the speed for the moment.

**Comment by rationalga** [comment]
Btw, When I enable  cuda acceleration, I get the following error. 
AttributeError: 'pyarrow.lib.RecordBatch' object has no attribute 'column_names'
without enabling cuda the training starts fine.
Detail error Below.
```

INFO:root:Randomly select 400 centroids from <lance.dataset.LanceDataset object at 0x7f78ac04d9d0>
INFO:root:Done sampling: centroids shape: torch.Size([400, 768])
INFO:root:Training IVF partitions using GPU(cuda)
INFO:root:Start kmean training, metric: cosine, iters: 50
  0%|                                                                                                                                                                               | 0/50 [00:00<?, ?it/s]INFO:root:Sampled at offset=23474, len=200
INFO:root:Sampled at offset=1778589, len=200
INFO:root:Sampled at offset=3567171, len=200
  0%|                                                                                                                                                                               | 0/50 [00:00<?, ?it/s]
Traceback (most recent call last):
build_index_from_hfds.py", line 162, in <module>
    table.create_index(metric="cosine", num_partitions=400, num_sub_vectors=8, accelerator="cuda")
    ../site-packages/lancedb/table.py", line 565, in create_index
    self._dataset.create_index(
   ../site-packages/lance/dataset.py", line 992, in create_index
    ivf_centroids = train_ivf_centroids_on_accelerator(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ../python3.11/site-packages/lance/vector.py", line 173, in train_ivf_centroids_on_accelerator
    kmeans.fit(ds)
../site-packages/lance/torch/kmeans.py", line 135, in fit
    self.total_distance = self._fit_once(
                          ^^^^^^^^^^^^^^^
../site-packages/lance/torch/kmeans.py", line 181, in _fit_once
    for idx, chunk in enumerate(data):
  ../site-packages/lance/torch/data.py", line 123, in __iter__
    yield _to_tensor(batch)
          ^^^^^^^^^^^^^^^^^
../site-packages/lance/torch/data.py", line 39, in _to_tensor
    for col in batch.column_names:
               ^^^^^^^^^^^^^^^^^^
AttributeError: 'pyarrow.lib.RecordBatch' object has no attribute 'column_names'

```

**Comment by eddyxu** [comment]
Could you check what version of pyarrow being used ?

**Comment by rationalga** [comment]
pyarrow==12.0.1

**Comment by chebbyChefNEQ** [comment]
ah, you need may need to upgrade to `pyarrow==14`

**Comment by rationalga** [comment]
Upgraded!. it works now. Thank you.

**Comment by chebbyChefNEQ** [comment]
another thing you need when using MPS or CUDA is to set multiprocessing start method like this
```python
from multiprocessing import set_start_method

if __name__ == "__main__":
    set_start_method("spawn")
```
this is required because we use `multiproessing` for loading data asynchronously and torch requires `spawn` instead of `fork`

**Comment by rationalga** [comment]
```
> RuntimeError: CUDA error: an illegal memory access was encountered
> another thing you need when using MPS or CUDA is to set multiprocessing start method like this
> 
> from multiprocessing import set_start_method
> 
> if __name__ == "__main__":
>     set_start_method("spawn")
> this is required because we use multiproessing for loading data asynchronously and torch requires spawn instead of fork. Otherwise the code will deadlock

```
Thank you. This was helpful as without this I got the following error.

RuntimeError: CUDA error: an illegal memory access was encountered
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.

**Comment by eddyxu** [comment]
Yea we should definitely improve our GPU training docs. cc @chebbyChefNEQ @QianZhu

**Comment by rationalga** [comment]
Hi, The code changes you made, are they already in the release version? (If I upgrade my pip install, would I be able to use the memory usage parameters?)
Thank you.

**Comment by chebbyChefNEQ** [comment]
Once indexed, you can query with `lancedb` as usual

**Comment by rationalga** [comment]
Hi,
Here are some observations.
It appears that choosing num_partitions has bigger impact even when shuffle_partition_batches is set to be very low. e.g. 
`
ds.create_index(column="vector", index_type="IVF_PQ", metric="cosine", num_partitions=16384, num_sub_vectors=64, shuffle_partition_batches = 10, shuffle_partition_concurrency=1)
`
here the RAM usages exceeds 32GB. 

For large datasets the number of centroids is  usually picked to be  quite large e.g.
```
 If N is between 10 Million and 100 Million
        num_partitions = 2**18
If N is between 100 Million and 1 Billion
        num_partitions = 2**20

```
I believe the value selected for num_partitions influences the number of embeddings sampled to create clusters. My question is, if there's a formula or method to predict memory usage based on these parameters. This would enable developers to select suitable parameters depending on their system's RAM.
Thank you.

**Comment by chebbyChefNEQ** [comment]
Hi there,

Usually when the number of partitions gets greater than 10k, training becomes very difficult and requires large GPUs. (50k+ requires distributed GPU training)

This is an inherent characteristic with the IVFPQ index. To combat this, I would suggest using much less partitions. The tradeoff is slower query performance and some query recall.

Could you try something like

num_partitions = num rows //[10000, 20000]
num_sub_vectors = dims // 16

In our experience this should be able to scale to 100M rows.
Say IVF8192PQ96


To scale even beyond 100M, we would love to have a quick chat and find out more about your use case in order to provide more guidance.

**Comment by rationalga** [comment]
Hi,
Very sorry  for not responding sooner. Due to some personal circumstances, I had to be away for a long while. I will try out your suggestion and get back to you with the results. Thank you.
